### PR TITLE
Pos tagging

### PR DESCRIPTION
**Pull request by jesseyli**

### CODE REVIEW

File: pos_blstm/blstm.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
+"""Implements the long-short term memory character model.
+This version vectorizes over multiple examples, but each string
+has a fixed length."""
```
Comment (mhw32) [inline pos_blstm/blstm.py]: What does this mean? Why do strings have "fixed length"?

File: pos_blstm/blstm.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+            'forget1':       rp(input_size + state_size + 1, state_size),
+            'ingate1':       rp(input_size + state_size + 1, state_size),
+            'outgate1':      rp(input_size + state_size + 1, state_size),
+            'init cells2':   rp(1, state_size),
+            'init hiddens2': rp(1, state_size),
+            'change2':       rp(input_size + state_size + 1, state_size),
+            'forget2':       rp(input_size + state_size + 1, state_size),
+            'ingate2':       rp(input_size + state_size + 1, state_size),
+            'outgate2':      rp(input_size + state_size + 1, state_size),
+            'predict':      rp(state_size*2 + 1, output_size)}
```
Comment (mhw32) [inline pos_blstm/blstm.py]: maybe call these ingate_f and ingate_b so we don't get confused which one is forward and which is backward.  Nice job adding the bias vectors (i usually forget those)

File: pos_blstm/blstm.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        change  = np.tanh(concat_and_multiply(params['change2'], _input, hiddens))
+        forget  = sigmoid(concat_and_multiply(params['forget2'], _input, hiddens))
+        ingate  = sigmoid(concat_and_multiply(params['ingate2'], _input, hiddens))
+        outgate = sigmoid(concat_and_multiply(params['outgate2'], _input, hiddens))
+        cells   = cells * forget + ingate * change
+        hiddens = outgate * np.tanh(cells)
+        return hiddens, cells
+
+    def hiddens_to_output_probs(hiddens):
+        output = concat_and_multiply(params['predict'], hiddens)
```
Comment (mhw32) [inline pos_blstm/blstm.py]: Are you sure the BiLSTM uses both forward and backward to produce outputs? 
As in: is the dimensionality of 'predict' : (state_size*2 + 1)x(output_size) or is it (state_size+1)x(output_size)?

I'm not positive but you might just use the forward pass to classify?

File: pos_blstm/blstm.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        # print((np.exp(output)/np.sum(np.exp(output), axis=1, keepdims=True))[0,:])
+        # print(output.shape)
+        return output - logsumexp(output, axis=1, keepdims=True) # Normalize log-probs.
+        # pdb.set_trace()
+        return np.exp(output)/np.sum(np.exp(output), axis=1, keepdims=True)
```
Comment (mhw32) [inline pos_blstm/blstm.py]: use line 67 -- keep tihngs in log space

File: pos_blstm/blstm.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    sentence = []
+    correct = []
+    for i in xrange(num_sentences):
+        for j in xrange(test_mask[i]):
+            sentence.append(one_hot[np.argmax(np.exp(logprobs)[:,i,:][j])])
+            correct.append(one_hot[np.argmax(Y_test[j,i,:])])
+    print(sum(np.array(sentence) == np.array(correct))/float(len(sentence)))
+
+    if save:
+        np.savez('tmp/trained_weights.npz', **trained_params)
```
Comment (mhw32) [inline pos_blstm/blstm.py]: Does it currently work? You should compare it with the standard example  from the LSTM file in github. This should do no worse. (https://github.com/HIPS/autograd/blob/master/examples/lstm.py)

File: pos_blstm/create_data.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
+from gensim import models
+import pickle
+import utils
+import pdb
+import numpy as np
+import pdb
+
+model = models.Word2Vec.load('tmp/embedding_map')
```
Comment (mhw32) [inline pos_blstm/create_data.py]: slowly i'd prefer to filter out gensim and nltk. Is this crucial?

File: pos_blstm/optimizers.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    v = np.zeros(len(x))
+    for i in range(num_iters):
+        g = flattened_grad(x, i)
+        if callback: callback(unflatten(x), i, unflatten(g))
+        m = (1 - b1) * g      + b1 * m  # First  moment estimate.
+        v = (1 - b2) * (g**2) + b2 * v  # Second moment estimate.
+        mhat = m / (1 - b1**(i + 1))    # Bias correction.
+        vhat = v / (1 - b2**(i + 1))
+        x = x - step_size*mhat/(np.sqrt(vhat) + eps)
+    return unflatten(x)
```
Comment (mhw32) [inline pos_blstm/optimizers.py]: Look in the disambiguator's optimizers. We shouldn't dup code + i have a batch-based ADAM. Batch SGD is faster!

File: pos_blstm/rnn.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+from autograd.scipy.misc import logsumexp
+from os.path import dirname, join
+from builtins import range
+from optimizers import adam
+
+
+### Helper functions #################
+
+def sigmoid(x):
+    return 0.5*(np.tanh(x) + 1.0)   # Output ranges from 0 to 1.
```
Comment (mhw32) [inline pos_blstm/rnn.py]: we should create a ml_utils.py file and share functions across these files

File: pos_blstm/blstm.py (hunk 1 above)
Comment (jesseyli) [reply pos_blstm/blstm.py]: This is getting removed, not related

File: pos_blstm/blstm.py (hunk 3 above)
Comment (jesseyli) [reply pos_blstm/blstm.py]: I'm not 100% sure about this but I think you need to feed both forward and backward to the output layer, if I only used to forward pass to classify then wouldn't it not consider any information about the reverse ordering?

File: pos_blstm/blstm.py (hunk 5 above)
Comment (jesseyli) [reply pos_blstm/blstm.py]: Right now it's getting accuracies in the low 80%'s, running it against the example LSTM is a good idea. I don't know what's causing the low accuracy yet.

File: pos_blstm/create_data.py (hunk 6 above)
Comment (jesseyli) [reply pos_blstm/create_data.py]: Definitely going to remove gensim later, but a good word embedding is going to necessary in order to get the best accuracy. Just using Word2Vec for now to build the BLSTM.

**Pull request merged**